SignalShareExtension/ShareViewController.swift

```diff
@@ -428,46 +498,103 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
         }
         Logger.info("\(self.logTag) attachment: \(itemProvider)")
 
-        // Order matters if we want to take advantage of share conversion in loadItem,
-        // Though currently we just use "data" for most things and rely on our SignalAttachment
-        // class to convert types for us.
-        let utiTypes: [String] = [kUTTypeImage as String,
-                                  kUTTypeURL as String,
-                                  kUTTypeData as String]
-
-        let matchingUtiType = utiTypes.first { (utiType: String) -> Bool in
-            itemProvider.hasItemConformingToTypeIdentifier(utiType)
-        }
-
-        guard let utiType = matchingUtiType else {
+        // We need to be very careful about which UTI type we use.
+        //
+        // * In the case of "textual" shares (e.g. web URLs and text snippets), we want to
+        //   coerce the UTI type to kUTTypeURL or kUTTypeText.
+        // * We want to treat shared files as file attachments.  Therefore we do not
+        //   want to treat file URLs like web URLs.
+        // * UTIs aren't very descriptive (there are far more MIME types than UTI types)
+        //   so in the case of file attachments we try to refine the attachment type
+        //   using the file extension.
+        guard let srcUtiType = ShareViewController.utiTypeForItem(itemProvider: itemProvider) else {
             let error = ShareViewControllerError.unsupportedMedia
             return Promise(error: error)
         }
-        Logger.debug("\(logTag) matched utiType: \(utiType)")
+        Logger.debug("\(logTag) matched utiType: \(srcUtiType)")
 
-        let (promise, fulfill, reject) = Promise<URL>.pending()
+        let (promise, fulfill, reject) = Promise<(URL, String)>.pending()
```

Comment:
TIOLI: Now that this method is longer and this promise returns multiple objects, it would help with readbility to return a named tuple.


e.g. 
```
let (promise, fulfill, reject) = Promise<(attachmentURL: URL, utiType: String)>.pending()
```

Reply:
Done.